### PR TITLE
Convert com.microsoft::ATen into org.pytorch.aten::ATen onnx op

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Pytorch ONNX exporter now emits org.pytorch.aten::ATen nodes with operator_name and overload_name attribute

### CODE REVIEW

**Comment by thiagocrepaldi** [comment]
@SherlockNoMad or @iK1D could you review this one? It is part of changes on pytorch exporter to support native aten fallback (without caffe2-specific stuff in it)

**Review by baijumeswani — DISMISSED** [review]
These changes look good to me. @iK1D please share your thoughts if anything seems to have been missed out.

**Comment by thiagocrepaldi** [comment]
@iK1D LMK if it is OK to merge this PR

**Comment by thiagocrepaldi** [comment]
@SherlockNoMad PR is approved by non-code owner. Could you help?

File: orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py
Comment (iK1D) [inline orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py:94]: Why do we have to add the overload_name when it's an optional attribute?

**Review by iK1D — COMMENTED** [review]
Maybe we can also add ATen to training_ops.cpu.json so the output warning from UT KernelDefHashTest.ExpectedCpuKernelDefHashes will not contain ATen op?

File: onnxruntime/core/providers/cpu/cpu_provider_shared.cc
Comment (thiagocrepaldi) [reply onnxruntime/core/providers/cpu/cpu_provider_shared.cc:169]: I was wondering the same, but as you mentioned, I will rename these occurences

**Comment by thiagocrepaldi** [comment]
> training_ops.cpu.json

@iK1D  How does this json work exactly? how do I get the hash for aten?

**Comment by thiagocrepaldi** [comment]
@snnn CI is failing with "unused variable" for a file that I didnt change. Is that a known CI issue?

https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=570910&view=logs&j=c65bf353-89c3-5a09-d766-68fb83c93231&t=19fced32-92b6-5d58-720f-91700d1ae625

**Comment by snnn** [comment]
Yes. Please wait them fixing it.

**Review by iK1D — APPROVED** [review]

**Review by baijumeswani — APPROVED** [review]

**Comment by thiagocrepaldi** [comment]
https://github.com/pytorch/pytorch/pull/68490 has been merged by PyTorch. This PR can not be merged as soon as CI allows it
@natke

**Review by baijumeswani — APPROVED** [review]

**Review by iK1D — APPROVED** [review]

**Comment by thiagocrepaldi** [comment]
@YUNQIUGUO @edgchen1 @gwang-msft @skottmckay could you review this PR? I already have approvals from ORT core, but none of them are code owners for this piece. Thanks

File: tools/python/util/ort_format_model/operator_type_usage_processors.py
Comment (edgchen1) [inline tools/python/util/ort_format_model/operator_type_usage_processors.py:28]: does it need to be added here?
see comment above: `# This list is limited to just the domains we have processors for`

**Comment by edgchen1** [comment]
> > training_ops.cpu.json
> 
> @iK1D How does this json work exactly? how do I get the hash for aten?

it contains kernel def hashes for ops enabled by the symbol ENABLE_TRAINING_OPS

some more info here:
https://github.com/microsoft/onnxruntime/blob/1cdc23aba419cab17aa9cc5faedfdd1bef8ab023/onnxruntime/test/providers/kernel_def_hash_test.cc#L5-L51

you can see how to dump the hashes there.

if and only if this op has a kernel that is enabled by ENABLE_TRAINING_OPS, add its kernel def hash to training_ops.cpu.json

File: tools/python/util/ort_format_model/operator_type_usage_processors.py
Comment (thiagocrepaldi) [reply tools/python/util/ort_format_model/operator_type_usage_processors.py:28]: TBH, I am not sure what the author meant with "processor" in this context :)

Comment (edgchen1) [reply tools/python/util/ort_format_model/operator_type_usage_processors.py:28]: I think it doesn't need to be added.
@skottmckay can you confirm?

Comment (thiagocrepaldi) [reply tools/python/util/ort_format_model/operator_type_usage_processors.py:28]: @skottmckay could you confirm whether adding `ort.pytorch.aten` domain to `operator_type_usage_processors.py` is required for replacing ORT's custom Aten OP by the recently introduced PyTorch's?

Comment (skottmckay) [reply tools/python/util/ort_format_model/operator_type_usage_processors.py:28]: This is used for determining the operators and types used by a model for creating a build with a reduced set of supported operators/types. I would not have expected pytorch ATen operators to be applicable to that scenario given the ATen library, IIRC, is massive, saving a few MB for the ORT build would be insignificant.

Comment (natke) [reply tools/python/util/ort_format_model/operator_type_usage_processors.py:28]: Sounds like we have the answer @thiagocrepaldi and @edgchen1

Comment (thiagocrepaldi) [reply tools/python/util/ort_format_model/operator_type_usage_processors.py:28]: @skottmckay ATEN library is not shipped in ORT nor ORT training wheel packages. Users are required to install pytorch as a dependency for training, which has the aten library there. So keeping or not this code would not imply any binary size increase. In that case, we could keep this line

However, if tools/python/util/ort_format_model/operator_type_usage_processors.py is meant for a ORT package that does not include ORTModule, then we could remove it

@skottmckay @edgchen1 What I do? Remove it or keep it?

Comment (natke) [reply tools/python/util/ort_format_model/operator_type_usage_processors.py]: What would happen in the scenario where there are aten ops in the model? 
- and we do keep this line?
- and we don't keep this line?

Comment (thiagocrepaldi) [reply tools/python/util/ort_format_model/operator_type_usage_processors.py]: AFAIK only ORT training supports aten ops

**Review by edgchen1 — DISMISSED** [review]

File: tools/python/util/ort_format_model/operator_type_usage_processors.py
Comment (skottmckay) [reply tools/python/util/ort_format_model/operator_type_usage_processors.py]: Inferencing definitely does NOT support aten ops.

This script is used to create the config for a reduced ops/types build. The context for wanting to do that is you need a smaller binary size for deployment to say a mobile device. I don't think that will ever be applicable to a training build that includes ATEN support.

It doesn't technically hurt to keep the line, but I think it's slightly better not to have it as you'll get an error if you're attempting to create a config based on a model with ATEN ops. We have never attempted or thought through the implications of doing a reduced ops build with that (or the sanity of even wanting to), so at least an error will alert you to someone trying to do that.

**Review by edgchen1 — APPROVED** [review]

**Comment by chilo-ms** [comment]
This PR is already in rel-1.11.0 branch, so remove the label.

**Pull request merged**
